### PR TITLE
kubeadm: Make the hostPath volume mount code more secure

### LINKED ISSUE #341: The hostPath volume mounting code is not optimal
It:
 - mounts everything as read-write
 - is not unit tested
 - is in a single, large function
 - is hard to read
 - mounts more than components need (e.g. `/etc/kubernetes` is mounted for the controller-manager (in the same dir as superuser privs, admin.conf, when it only needs `controller-manager.conf`)

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

 - Refactors the hostpath volume mounting code for the Static Pods
 - Splits out the functionality that was in a big function to something testable
 - Unit test a lot
 - Adds support for mounting external etcd CA/cert/key files in an other path than `/etc/ssl/certs`. Before this you **had** to have your files in there or the apiserver would crashloop
 - Significantly improves comment coverage
 - Now only mounts the bare essentials instead of nearly everything. For example, don't mount full `/etc/kubernetes` when the only thing you need is `/etc/kubernetes/scheduler.conf`
 - Make everything but the etcd datadir read-only for components.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Fixes: https://github.com/kubernetes/kubeadm/issues/341

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```
cc @kubernetes/sig-cluster-lifecycle-pr-reviews

### CODE REVIEW

**Comment by luxas** [comment]
@timothysc PTAL

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -50,74 +50,71 @@ const (
...
+
 	// Prepare static pod specs
 	staticPodSpecs := map[string]v1.Pod{
 		kubeAPIServer: componentPod(v1.Container{
 			Name:          kubeAPIServer,
 			Image:         images.GetCoreImage(images.KubeAPIServerImage, cfg, kubeadmapi.GlobalEnvParams.HyperkubeImage),
-			Command:       getAPIServerCommand(cfg, false, k8sVersion),
-			VolumeMounts:  volumeMounts,
+			Command:       getAPIServerCommand(cfg, k8sVersion),
+			VolumeMounts:  volumeMounts[kubeAPIServer],
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: The old schooler in me says you might want to hide this detail behind some utility functions, but not a concern at this time.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -146,6 +143,71 @@ func WriteStaticPodManifests(cfg *kubeadmapi.MasterConfiguration) error {
...
+			// Mount the public CA certs
+			caCertsVolume,
+		},
+		kubeControllerManager: {
+			// Mount the required certificates
+			k8sCertsVolume,
+			// Mount the public CA certs
+			caCertsVolume,
+			// Mount the kubeconfig file for talking to the API server
+			newVolume(kubeConfigVolumeName, filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ControllerManagerKubeConfigFileName)),
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Why is this not RO?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 3]
```diff
@@ -146,6 +143,71 @@ func WriteStaticPodManifests(cfg *kubeadmapi.MasterConfiguration) error {
 	return nil
 }
 
+// getHostPathVolumesForTheControlPlane gets the volumes and volumemounts needed for each control plane component
+func getHostPathVolumesForTheControlPlane(cfg *kubeadmapi.MasterConfiguration) (map[string][]v1.Volume, map[string][]v1.VolumeMount) {
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: I'd consider just create a struct that hides the details, this function is a little hard to read.  Someone not intimately familiar with what is going here would probably get confused.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 4]
```diff
@@ -155,97 +217,73 @@ func newVolume(name, path string) v1.Volume {
...
-		Name: "etcd",
-		VolumeSource: v1.VolumeSource{
-			HostPath: &v1.HostPathVolumeSource{Path: cfg.Etcd.DataDir},
-		},
-	}
-}
-
-func etcdVolumeMount(dataDir string) v1.VolumeMount {
+// newROVolumeMount creates a read-only v1.VolumeMount to the specified location
+func newROVolumeMount(name, path string) v1.VolumeMount {
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: could dedupe these two to call a single function that passes RO=true|false

**Review by timothysc — COMMENTED** [review]
I don't really have strong opinions on my comments.  I do think it's hard to follow some of the logic around volume and volumeMnts.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: I'll try to make this a little easier to read then, that was part of my goal. I'll add a struct for it

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Volumes can't be RO, VolumeMounts can

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 4 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Yeah, I know, but I felt it would be easier to read RO/RW in func name vs passing a bool. Such a small duplication so I thought it was worth it.

**Comment by luxas** [comment]
@timothysc This is ready for re-review. I changed the implementation to address your comments, but didn't touch the unit tests that just passed.

I also splitted things out to two files instead of one mega-file.

**Review by timothysc — APPROVED** [review]
/lgtm 

I'm ok with it, but your defer shuffle on shared variables is problematic.
